Use uppercase as additional characteristic for concept linking

myrthemh opened the pull request:
Hi @zeljko, the previous version of MedCAT contained the property "name_isupper" to track whether a concept in the concept table is completely in uppercase. In that previous version, Sander Tan @sandertan (data scientist at UMC Utrecht) and I (intern at UMC Utrecht) made a modification to MedCAT to use this property during entity linking. This is useful for ambiguous terms that are both a medical acronym and a non-medical word. For example, we want to distinguish ALS (Amyotrophic lateral sclerosis) from the Dutch word "als", which is a conjunction in Dutch. For English words, the same problem happens for terms such as MS (Multiple Sclerosis), FISH (fluorescence in situ hybridization), MAP (Microtubule-Associated Proteins), and some gene names. It wasn't possible to solve this with MedCAT disambiguation functionality because then we would have to add non-medical words to the concept table.
 
This PR suggests to add the boolean property "is_upper" when adding concepts to the concept database, and use this property during linking. Note that this will change the CDB model, so to use this new functionality a new CDB should be made.
 
What do you think of adding this functionality to MedCAT? Do the current changes suffice, or do you prefer a different approach?

w-is-h commented:
Hi @myrthemh,

sorry for removing that, and no problem we can add it back.

Re the pull request, we just have to sort out a couple of things. 

Do you want to check uppercase match only for **one** token as this line will only do it for one and in case of multiple tokens it will be ignored` if name in self.cdb.name2cuis and not tkn.is_stop and tkn.is_upper == self.cdb.name_isupper.get(name, False):`

You've also added ` Token.set_extension('is_upper', default=False, force=True)`  isn't this a duplicate of `token.is_upper`?

In the cdb_maker when setting isupper [here](https://github.com/CogStack/MedCAT/blob/4d22f22fb0e9ebfa95382ae6e1164a10babc3bec/medcat/cdb_maker.py#L159) it seems that you are applying the same `isupper` to all different names that can be found in one CSV row. This could be problematic as different names could have different casing. I suggest you edit the prepare_name function from cleaners.

And finally, it would e good if this was configurable, so that in the config you can set do you want to match case or not.

sandertan commented:
@myrthemh perhaps we can move the triple:
```
if self.config.general['case_sensitive']:
                     if tkn.is_upper == self.cdb.name_isupper.get(name, False):
```
to a single one in `def maybe_annotate_name()`

Also, there's a double `the` in `# Additional checking whether the the concepts...`

Lastly, I'm reconsidering the name of the property in the config. `case_sensitive` might imply a more broad functionality for case sensitive matching. Perhaps `check_upper_case_names` is more appropriate here. I named it specifically _names_ instead of _concepts_ because the check is for the names of the concepts. @w-is-h what do you think?

w-is-h commented:
We are on the right track, @sandertan `check_upper_case_names` looks good. Moving the check for uppercase inside of the `maybe_annotate` is also a good idea, as then we do not have to duplicate the code. 

One more simplification:
```
        is_upper = None

        if config.general['case_sensitive']:
            is_upper = sc_name.text.isupper()
```
There is no need to check this in `cleaners.py` you can just say `is_upper = sc_name.text.isupper()` always.

myrthemh commented:
Thank you for your feedback. I have simplified the code in cleaners, renamed the config variable `case-sensitive` to `check_upper_case_names`, removed the double `the` of the comment `# Additional checking whether the the concepts...` and moved the triple 
```
if self.config.general['case_sensitive']:
                     if tkn.is_upper == self.cdb.name_isupper.get(name, False):
```
to a single one in `def maybe_annotate_name()`. 

In addition, I slightly changed the two statements above to one statement and improved checking the names that consist of more than one token with the code: 

 ```
    if not config.general['check_upper_case_names'] or (config.general['check_upper_case_names'] and all([x.is_upper for x in tkns]) == cdb.name_isupper.get(name, False)):

```

sandertan commented:
@w-is-h We rebased the branch on current master and added your suggestions in the last commit. What do you think?

w-is-h commented:
@sandertan false alarm, all good, merged.

sandertan commented:
Fantastic, thanks!
